### PR TITLE
[FLINK-3952][runtine] Upgrade to Netty 4.1

### PR DESCRIPTION
**Pull request by pnowojski**
This PR bases on https://github.com/apache/flink/pull/6031 and first two commits should be ignored here.

This PR adjusts our code to work with Netty 4.1. It also includes possible bug fix to file uploading cleanup in FileUploadHandler and HttpRequestHandler. For mor information look here:

https://github.com/netty/netty/issues/7611

`Embed flink-shaded-netty-4` commit is only for having green travis and will be dropped once new `flink-shadded-netty` will be released.

## Verifying this change

This change is covered by variety of pre existing tests. Furthermore I have manually verified that issue mentioned by @uce in the commit message here: https://github.com/apache/flink/commit/d92e422ec7089376583a8f57043274d236c340a4
doesn't happen: 
- I have reproduced this issue on a test cluster with Flink 1.0-XXX
- I have verified that the same job passes without any problems after upgrading to Netty 4.1

I have also run our network benchmark suite and verified that there are no performance changes after this change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -308,7 +308,7 @@ under the License.
 				errors.
 
 				[1] https://github.com/netty/netty/issues/3704 -->
```
Comment (StephanEwen) [inline pom.xml]: Looks like you can remove this comment now.

Comment (pnowojski) [reply pom.xml]: Ops, dropped.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java [hunk 2]
```diff
@@ -247,7 +249,15 @@ public void shutdown(Time timeout) {
 
 		@Override
 		protected void channelRead0(ChannelHandlerContext ctx, Object msg) {
-			if (msg instanceof FullHttpResponse) {
+			// TODO: should this check for status OK (200) and treat all other as errors?
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: message semantics are handled later when parsing the payload, which effectively does what you're suggesting. The else branch for debugging; our rest servers always return `FullHttpResponses`.

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: Thx for filling this out. In that case should I just drop this `// TODO:`?

**Comment by pnowojski** [comment]
Dead lock visible in failed travis is unrelated to this change (https://issues.apache.org/jira/browse/FLINK-9439)

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: yes I think we can just drop it.

**Comment by pnowojski** [comment]
@zentol I have also fixed added a test to cover https://github.com/apache/flink-shaded/issues/30 in the second commit

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RoutedRequest.java [hunk 3]
```diff
@@ -94,4 +94,14 @@ public ReferenceCounted retain(int arg0) {
...
 	}
+
+	@Override
+	public ReferenceCounted touch() {
+		return this;
+	}
+
+	@Override
+	public ReferenceCounted touch(Object o) {
+		return this;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RoutedRequest.java]: We may want to `touch` the contained request as well like in [this](https://github.com/netty/netty/blob/4.1/codec-redis/src/main/java/io/netty/handler/codec/redis/ArrayRedisMessage.java) class, depending on `requestAsReferenceCounted`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java [hunk 4]
```diff
@@ -75,12 +75,12 @@ private ReadOnlySlicedNetworkBuffer(ByteBuf buffer, int index, int length, int m
 
 	@Override
 	public ByteBuf unwrap() {
-		return super.unwrap().unwrap();
+		return super.unwrap();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java]: Is this change necessary for the upgrade or cleanup?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBuffer.java [hunk 5]
```diff
@@ -215,16 +242,33 @@ protected void _setMedium(int index, int value) {
 		setByte(index + 2, (byte) value);
 	}
 
+	@Override
+	protected void _setMediumLE(int index, int value){
+		setByte(index, (byte) value);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBuffer.java]: was this also taken from `UnpooledDirectByteBuf`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBuffer.java [hunk 6]
```diff
@@ -424,6 +480,19 @@ public int setBytes(int index, ScatteringByteChannel in, int length) throws IOEx
...
 	}
 
+	@Override
+	public int setBytes(int index, FileChannel in, long position, int length) throws IOException {
+		// adapted from UnpooledDirectByteBuf:
+		checkIndex(index, length);
+
+		ByteBuffer tmpBuf = memorySegment.wrap(index, length);
+		try {
+			return in.read(tmpBuf);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBuffer.java]: are you ignoring the `position` argument intentionally?

File: flink-tests/src/test/java/org/apache/flink/test/runtime/NettyEpollITCase.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+					TASK_MANAGERES,
+					1),
+				true);
+			cluster.before();
+			return Optional.of(cluster);
+		}
+		catch (UnsatisfiedLinkError ex) {
+			// If we failed to init netty because we are not on Linux platform, abort the test.
+			if (findThrowableWithMessage(ex, "Only supported on Linux").isPresent()) {
+				return Optional.empty();
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/NettyEpollITCase.java]: I didn't know about `AssumptionViolatedException`. Thanks for pointing this out.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBuffer.java (hunk 6 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBuffer.java]: ops, good catch. Even better catch is that this was not covered by any test, because I assumed `AbstractByteBufTest` comes from netty, while in reality it was copied to our code.

Fixed this lack of `position` bug and upgraded the `AbstractByteBufTest` as well - newer version was correctly failing with this bug.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java (hunk 4 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java]: Necessary for the upgrade, otherwise our tests (including `ITCases`) are failing. In upgraded version there is a different structure of nested classes in case of using slices of readonly buffers. Possibly this is a bug fix for a bug, that didn't show up in older netty.

File: pom.xml [hunk 8]
```diff
@@ -300,15 +300,7 @@ under the License.
...
-				<!-- Don't upgrade for now. Netty versions >= 4.0.28.Final
-				contain an improvement by Netty, which slices a Netty buffer
-				instead of doing a memory copy [1] in the
-				LengthFieldBasedFrameDecoder. In some situations, this
-				interacts badly with our Netty pipeline leading to OutOfMemory
-				errors.
-
-				[1] https://github.com/netty/netty/issues/3704 -->
-				<version>4.0.27.Final-${flink.shaded.version}</version>
+				<version>4.1.24.Final-${flink.shaded.version}</version>
```
[the remark this answers is not in the document]
Comment (zentol) [reply pom.xml]: I've opened #6128 to bump the remaining version so this PR can stay focused on netty.

**Comment by pnowojski** [comment]
@zentol as @NicoK suggested I have missed netty 4.1 upgrade without credit base flow control. It turns out that this upgrade is unstable without this fix: https://github.com/apache/flink/pull/5571, so we should merge it before this one.

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java [hunk 9]
```diff
@@ -88,6 +102,16 @@ public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
 		}
 	}
 
+	@Override
+	public long length() {
+		return length;
```
Comment (tillrohrmann) [inline flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java]: Is it always guaranteed that `buf.readableBytes` won't change? If not, then I think we should return `-1` here, because the length is unknown.

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java [hunk 10]
```diff
@@ -88,6 +102,16 @@ public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
...
+	}
+
+	@Override
+	public long progress() {
+		return length - buf.readableBytes();
```
Comment (tillrohrmann) [inline flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java]: If the length is not know, then I think it would be better to return something like `buf.readerIndex - initialReaderIndex`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java]: But isn't it a bit counter-intuitive that you instantiate a `ReadOnlySlicedNetworkBuffer` with a `NetworkBuffer` and when you call `ReadOnlySlicedNetworkBuffer.unwrap` you don't get a `NetworkBuffer` back but a `SlicedByteBuf`? Thus, you need to know the internals of this class to know that you have to call twice `unwrap` to obtain the `NetworkBuffer`. This shows for example in the `NetworkBufferTest` where we always call `slice.unwrap().unwrap()`.

What exactly did change with the structure of nested classes that this change is necessary?

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/AbstractByteBufTest.java [hunk 11]
```diff
@@ -64,12 +69,13 @@
...
+import static org.junit.Assume.assumeFalse;
 import static org.junit.Assume.assumeTrue;
 
 /**
  * An abstract test class for channel buffers.
  *
- * Copied from netty 4.0.50 with some changes to fit our netty version 4.0.27.
+ * Copy from netty 4.1.24.Final.
  */
 public abstract class AbstractByteBufTest {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/AbstractByteBufTest.java]: Let's extend this class from `TestLogger`.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/NettyEpollITCase.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+			input.keyBy(new KeySelector<Integer, Integer>() {
+					@Override
+					public Integer getKey(Integer value) throws Exception {
+						return value;
+					}
+				})
+				.sum(0)
+				.print();
+
+			env.execute();
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/NettyEpollITCase.java]: Is it strictly necessary that this test must be an integration test case? As far as I understand the test case, wouldn't it be enough to setup a `NettyServer` with `epoll` activated and sending some data to it via a `NettyClient` with `epoll` activated? Since we don't assert anything else than whether the Flink program executes or not, it should be basically the same.

**Comment by tillrohrmann** [comment]
Changes look good to me @pnowojski. I had some minor comments. After addressing them, we should rebase and remove the unnecessary commits. Ping me once you have done it. I will then merge this PR.

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java (hunk 9 above)
Comment (pnowojski) [reply flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java]: It seems like content of `buf` is not changing. However we can not guard against it programmatically, only via java doc. On the other hand, I'm not sure what would the side effects of "unknown" length be.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RoutedRequest.java [hunk 13]
```diff
@@ -97,11 +98,17 @@ public ReferenceCounted retain(int arg0) {
...
+		if (requestAsReferenceCounted.isPresent()) {
+			ReferenceCountUtil.touch(requestAsReferenceCounted.get());
+		}
 		return this;
 	}
 
 	@Override
 	public ReferenceCounted touch(Object o) {
+		if (requestAsReferenceCounted.isPresent()) {
+			ReferenceCountUtil.touch(requestAsReferenceCounted.get());
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RoutedRequest.java]: I think we could do that. I'm not entirely sure how this whole `touch` is suppose to work, but passing down the hint shouldn't harm anything and be more correct.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/NettyEpollITCase.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/NettyEpollITCase.java]: The same question would remain: whether such artificial setup, setups a network stack in the same way as Flink.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java (hunk 4 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/ReadOnlySlicedNetworkBuffer.java]: as discussed offline, this is because of change in the implementation of `SlicedByteBuf`.

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java (hunk 9 above)
Comment (pnowojski) [reply flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java]: changed to `-1`

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java (hunk 10 above)
Comment (pnowojski) [reply flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/network/ChunkedByteBuf.java]: changed to `return buf.readerIndex()` since progress is not well defined.

**Comment by tillrohrmann** [comment]
Thanks for your work and addressing my comments @pnowojski. Travis gave green light. Merging this PR now.

**Comment by pnowojski** [comment]
Thanks!
